flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -195,9 +195,19 @@ void announceCombinedWatermark() {
                 "Distributing maxAllowedWatermark={} to subTaskIds={}",
                 maxAllowedWatermark,
                 subTaskIds);
-        for (Integer subtaskId : subTaskIds) {
-            context.sendEventToSourceOperator(
-                    subtaskId, new WatermarkAlignmentEvent(maxAllowedWatermark));
+        // Because of Java-ThreadPoolExecutor will not schedule the period task
+        // if it throws an exception, so we should handle the potential exception like
+        // "subtask xx is not ready yet to receive events" to increase robustness.
+        try {
+            for (Integer subtaskId : subTaskIds) {
+                context.sendEventToSourceOperator(
+                        subtaskId, new WatermarkAlignmentEvent(maxAllowedWatermark));
+            }
```

Comment:
If any subtask throws exception, the next subtask won't send event, right?

Reply:
yeah, just wait next time to send
in my opinion, if one task is during failing, other tasks maybe also during failing with high probability

Reply:
Hi @LoveHeat , thanks for your feedback.

As I said at FLINK-32362 before, I'm not sure what should we do when some subtasks are not ready.

- Option1: Send event to all ready subtasks, and just ignore unready subtasks.
- Option2: Don't send any event before all subtasks are ready.

If we expect option1, we should ensure the event is sent to all ready subtasks.

> In my opinion, if one task is during failing, other tasks maybe also during failing with high probability

When `jobmanager.execution.failover-strategy` is region, and all subtasks don't have shuffle link. If subtask0  cannot start or fails, other subtasks should work well, right?

Reply:
After thinking about it, I think it can work. When one task is failing, the aggWM won't be updated with high probability. So we can don't send events. 

I hope to hear more from @pnowojski 

BTW, I doubt the watermark alignment doesn't work well after a subtask is finished, because the finished subtaskIndex cannot be removed from the `subTaskIds`.

Reply:
Well, in the current diff, this logic may partially send events. If we want to make the entire operation atomic, we can check `context#registeredReaders()` or `SubtaskGateway` to check that all readers are available.

Reply:
After review code, partially-send-events is just ok, no negative impact(or i haven't found).
There are two cases that gateway is not ready: 
  1. task during deploying  
  2. task during restore before re-deployed (see code: [schedulerBase::notifyCoordinatorsOfSubtaskRestore](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java#L408) )

for case1:  it only limit the ready task consume speed to avoid it consume too faster than unready task, i think this is acceptable.
for case2: partially sending events have no impact on case2, because all task will be re-deployed.

@mas-chen

Reply:
Huh. I have a couple of remarks here:
- logging warn here, for an expected case that some subtask might not be yet running, is not ideal. But I'm not sure how to best handle that. Maybe only start printing it after a 5th consecutive failure? Can we somehow determine if an exception from sending an event is intermittent or not?
- should we actually fail the job at some point due to this kind of errors?
- I think we need a safety net when scheduling something in the `coordinatorExecutor`. This independent of properly handling the exception in this ticket, just to make sure in the future we won't have a similar bug due to unnoticed and unhandled exceptions. I would either:
  - remove `context.getCoordinatorExecutor()` getter, and allow to schedule runnables only through the context, like `SourceCoordinatorContext#runInCoordinatorThread`
  - or replace the returned executor with some wrapper that would actually fail the job on all unexpected failures

Reply:
How about this: remove `context.getCoordinatorExecutor() `, and provide a function to schedule period task like below, PeriodTaskHook is a hook to let caller decide whether ignore the failure or fail job.



     interface PeriodTaskHook {
       
       static PeriodTaskHook DO_NOTHING_HOOK = new PeriodTaskHook() {};

        default void taskFailure(Throwable t, SourceCoordinatorContext<?> context) {}

        default void taskSuccess() {}
    }

    ScheduledFuture<?> schedulePeriodTask(
            Runnable command,
            long initDelay,
            long period,
            TimeUnit unit,
            PeriodTaskHook taskHook) {
        return coordinatorExecutor.scheduleAtFixedRate(
                () -> {
                    try {
                        command.run();
                        taskHook.taskSuccess();
                    } catch (Throwable t) {
                        taskHook.taskFailure(t, this);
                    }
                },
                initDelay,
                period,
                unit);
    }


For SourceAlignment, its auto recoverable after task become ready, so we can ignore these failure? @pnowojski

Reply:
> remove context.getCoordinatorExecutor() , and provide a function to schedule period task like below, PeriodTaskHook is a hook to let caller decide whether ignore the failure or fail job.

Yes, I think we need something like that.

> For SourceAlignment, its auto recoverable after task become ready, so we can ignore these failure?

Ok, that should be fair enough, but we would need to distinguish those "subtask is not ready" errors from anything else. Ignoring all `Throwable ignored` is a bad idea, there might be other bugs or issues (`OOM`, `NPE`, and many many more). We can ignore "subtask is not ready", but we should fail the job on anything else.

Reply:
code is updated~ @pnowojski 
For now, the only RetryableErrors is taskNotReady, so i change the function name as sendEventToSourceOperatorIfTaskReady